In [11]:
import numpy as np
import pandas as pd

In [12]:
df= pd.read_csv('F://melbourne.csv')

In [13]:
df

Suburb              Address  Rooms Type      Price Method  \
0          Abbotsford        68 Studley St      2    h        NaN     SS   
1          Abbotsford         85 Turner St      2    h  1480000.0      S   
2          Abbotsford      25 Bloomburg St      2    h  1035000.0      S   
3          Abbotsford   18/659 Victoria St      3    u        NaN     VB   
4          Abbotsford         5 Charles St      3    h  1465000.0     SP   
5          Abbotsford     40 Federation La      3    h   850000.0     PI   
6          Abbotsford          55a Park St      4    h  1600000.0     VB   
7          Abbotsford         16 Maugie St      4    h        NaN     SN   
8          Abbotsford         53 Turner St      2    h        NaN      S   
9          Abbotsford         99 Turner St      2    h        NaN      S   
10         Abbotsford       129 Charles St      2    h   941000.0      S   
11         Abbotsford         124 Yarra St      3    h  1876000.0      S   
12         Abbotsford  121/56 Nicholson St      2    u        NaN     PI   
13         Abbotsford        17 Raphael St      4    h        NaN      W   
14         Abbotsford        98 Charles St      2    h  1636000.0      S   
15         Abbotsford     217 Langridge St      3    h  1000000.0      S   
16         Abbotsford      18a Mollison St      2    t   745000.0      S   
17         Abbotsford   6/241 Nicholson St      1    u   300000.0      S   
18         Abbotsford        10 Valiant St      2    h  1097000.0      S   
19         Abbotsford  403/609 Victoria St      2    u   542000.0      S   
20         Abbotsford            2 Rich St      2    h        NaN     SP   
21         Abbotsford    25/84 Trenerry Cr      2    u   760000.0     SP   
22         Abbotsford    106/119 Turner St      1    u   481000.0     SP   
23         Abbotsford   411/8 Grosvenor St      2    u   700000.0     VB   
24         Abbotsford      40 Nicholson St      3    h  1350000.0     VB   
25         Abbotsford  123/56 Nicholson St      2    u   750000.0      S   
26         Abbotsford           22 Park St      4    h  1985000.0      S   
27         Abbotsford    13/84 Trenerry Cr      1    u   500000.0      S   
28         Abbotsford        45 William St      2    h  1172500.0      S   
29         Abbotsford   7/20 Abbotsford St      1    u   441000.0     SP   
...               ...                  ...    ...  ...        ...    ...   
23517   Vermont South        2 Settlers Ct      4    h  1295000.0     SP   
23518   Vermont South        9 Winswood Cl      4    h  1250000.0     PI   
23519        Viewbank        149 Graham Rd      5    h  1316000.0     SP   
23520        Viewbank       7 Northwood Dr      4    h        NaN     SP   
23521        Wantirna   16 chesterfield Ct      4    h   951000.0      S   
23522        Wantirna            9 eton Sq      4    h        NaN     PI   
23523  Wantirna South         12 Armagh Cr      4    h  1323000.0      S   
23524  Wantirna South        34 Fewster Dr      3    h   970000.0      S   
23525  Wantirna South           15 Mara Cl      4    h  1330000.0      S   
23526        Watsonia       2/14 Daours Ct      2    u   590000.0      S   
23527        Watsonia        76 Kenmare St      2    h   650000.0     PI   
23528  Watsonia North         22 Meakin St      3    h        NaN     PI   
23529        Werribee          6 Feijoa Ct      5    h        NaN      W   
23530        Werribee       11 Gundaroo Sq      4    h        NaN     PI   
23531        Werribee         91 Latham St      4    h   540000.0     PI   
23532        Werribee          5 Nuragi Ct      4    h   635000.0      S   
23533        Werribee             3 Sky Ct      3    h        NaN     PI   
23534  West Footscray      1/21 Waratah St      2    t   655000.0      S   
23535     Westmeadows           9 Black St      3    h   582000.0      S   
23536     Westmeadows      81 Hillcrest Dr      4    h   720000.0     VB   
23537   Wheelers Hill         12 Strada Cr      4    h  1245000

In [14]:
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [15]:
df.tail()

Suburb           Address  Rooms Type      Price Method  \
23542  Wyndham Vale   25 Clitheroe Dr      3    u        NaN     PN   
23543  Wyndham Vale  19 Dalrymple Bvd      4    h        NaN      S   
23544     Yallambie      17 Amaroo Wy      4    h  1100000.0      S   
23545    Yarraville        6 Agnes St      4    h  1285000.0     SP   
23546    Yarraville     33 Freeman St      4    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
23542      Harcourts  26-08-2017      27.2    3024.0  ...       1.0  0.0   
23543  hockingstuart  26-08-2017      27.2    3024.0  ...       NaN  NaN   
23544     Buckingham  26-08-2017      12.7    3085.0  ...       3.0  2.0   
23545        Village  26-08-2017       6.3    3013.0  ...       1.0  1.0   
23546        Village  26-08-2017       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
23542     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23543       NaN           NaN        NaN          NaN -37.87882   144.60184   
23544       NaN           NaN        NaN          NaN -37.72006   145.10547   
23545     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23546       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
23542   Western Metropolitan        5262.0  
23543   Western Metropolitan        5262.0  
23544  Northern Metropolitan        1369.0  
23545   Western Metropolitan        6543.0  
23546   Western Metropolitan        6543.0  

[5 rows x 21 columns]

In [16]:
df.shape

(23547, 21)

In [17]:
df.keys()

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [18]:
df.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [19]:
#indexing and slicing

In [20]:
df[0,0]

KeyError: (0, 0)

In [21]:
df.iloc[0]

Suburb                      Abbotsford
Address                  68 Studley St
Rooms                                2
Type                                 h
Price                              NaN
Method                              SS
SellerG                         Jellis
Date                        03-09-2016
Distance                           2.5
Postcode                          3067
Bedroom2                             2
Bathroom                             1
Car                                  1
Landsize                           126
BuildingArea                       NaN
YearBuilt                          NaN
CouncilArea                      Yarra
Lattitude                     -37.8014
Longtitude                     144.996
Regionname       Northern Metropolitan
Propertycount                     4019
Name: 0, dtype: object

In [22]:
df.iloc[0:5]

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [23]:
df.iloc[:,'Rooms']

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

In [24]:
df.loc[:,'Rooms']

0        2
1        2
2        2
3        3
4        3
5        3
6        4
7        4
8        2
9        2
10       2
11       3
12       2
13       4
14       2
15       3
16       2
17       1
18       2
19       2
20       2
21       2
22       1
23       2
24       3
25       2
26       4
27       1
28       2
29       1
        ..
23517    4
23518    4
23519    5
23520    4
23521    4
23522    4
23523    4
23524    3
23525    4
23526    2
23527    2
23528    3
23529    5
23530    4
23531    4
23532    4
23533    3
23534    2
23535    3
23536    4
23537    4
23538    3
23539    3
23540    2
23541    4
23542    3
23543    4
23544    4
23545    4
23546    4
Name: Rooms, Length: 23547, dtype: int64

In [25]:
df.info()
#here in price---> it has  18396, it means there around 5k missing data..etc..

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
Suburb           23547 non-null object
Address          23547 non-null object
Rooms            23547 non-null int64
Type             23547 non-null object
Price            18396 non-null float64
Method           23547 non-null object
SellerG          23547 non-null object
Date             23547 non-null object
Distance         23546 non-null float64
Postcode         23546 non-null float64
Bedroom2         19066 non-null float64
Bathroom         19063 non-null float64
Car              18921 non-null float64
Landsize         17410 non-null float64
BuildingArea     10018 non-null float64
YearBuilt        11540 non-null float64
CouncilArea      15656 non-null object
Lattitude        19243 non-null float64
Longtitude       19243 non-null float64
Regionname       23546 non-null object
Propertycount    23546 non-null float64
dtypes: float64(12), int64(1), object(8)
memory usage: 3.8+ M

In [26]:
#indentifying the missing values

In [27]:
df.isnull()

Suburb  Address  Rooms   Type  Price  Method  SellerG   Date  Distance  \
0       False    False  False  False   True   False    False  False     False   
1       False    False  False  False  False   False    False  False     False   
2       False    False  False  False  False   False    False  False     False   
3       False    False  False  False   True   False    False  False     False   
4       False    False  False  False  False   False    False  False     False   
5       False    False  False  False  False   False    False  False     False   
6       False    False  False  False  False   False    False  False     False   
7       False    False  False  False   True   False    False  False     False   
8       False    False  False  False   True   False    False  False     False   
9       False    False  False  False   True   False    False  False     False   
10      False    False  False  False  False   False    False  False     False   
11      False    False  False  False  False   False    False  False     False   
12      False    False  False  False   True   False    False  False     False   
13      False    False  False  False   True   False    False  False     False   
14      False    False  False  False  False   False    False  False     False   
15      False    False  False  False  False   False    False  False     False   
16      False    False  False  False  False   False    False  False     False   
17      False    False  False  False  False   False    False  False     False   
18      False    False  False  False  False   False    False  False     False   
19      False    False  False  False  False   False    False  False     False   
20      False    False  False  False   True   False    False  False     False   
21      False    False  False  False  False   False    False  False     False   
22      False    False  False  False  False   False    False  False     False   
23      False    False  False  False  False   False    False  False     False   
24      False    False  False  False  False   False    False  False     False   
25      False    False  False  False  False   False    False  False     False   
26      False    False  False  False  False   False    False  False     False   
27      False    False  False  False  False   False    False  False     False   
28      False    False  False  False  False   False    False  False     False   
29      False    False  False  False  False   False    False  False     False   
...       ...      ...    ...    ...    ...     ...      ...    ...       ...   
23517   False    False  False  False  False   False    False  False     False   
23518   False    False  False  False  False   False    False  False     False   
23519   False    False  False  False  False   False    False  False     False   
23520   False    False  False  False   True   False    False  False     False   
23521   False    False  False  False  False   False    False  False     False   
23522   False    False  False  False   True   False    False  False     False   
23523   False    False  False  False  False   False    False  False     False   
23524   False    False  False  False  False   False    False  False     False   
23525   False    False  False  False  False   False    False  False     False   
23526   False    False  False  False  False   False    False  False     False   
23527   False    False  False  False  False   False    False  False     False   
23528   False    False  False  False   True   False    False  False     False   
23529   False    False  False  False   True   False    False  False     False   
23530   False    False  False  False   True   False    False  False     False   
23531   False    False  False  False  False   False    False  False     False   
23532   False    False  False  False  False   False    False  False     False   
23533   False    False  False  False   True   False    False  False     False   
23534   False    False  False  False  

In [28]:
#summing up the missing values (col-wise)
df.isnull().sum()
#in each column how many missing data is there inn col wise

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

In [29]:
#col having at least one missing value
df.isnull().any()

Suburb           False
Address          False
Rooms            False
Type             False
Price             True
Method           False
SellerG          False
Date             False
Distance          True
Postcode          True
Bedroom2          True
Bathroom          True
Car               True
Landsize          True
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude         True
Longtitude        True
Regionname        True
Propertycount     True
dtype: bool

In [30]:
#above is equivalent to axis=0(by default, any()--> operators on col)
df.isnull().any(axis=0)
#this is finding missing value in col-wise--->coz axis=0

Suburb           False
Address          False
Rooms            False
Type             False
Price             True
Method           False
SellerG          False
Date             False
Distance          True
Postcode          True
Bedroom2          True
Bathroom          True
Car               True
Landsize          True
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude         True
Longtitude        True
Regionname        True
Propertycount     True
dtype: bool

In [31]:
df.isnull().any(axis=1)
#this is finding missing value in row-wise--->coz axis=1

0         True
1         True
2        False
3         True
4        False
5         True
6        False
7         True
8         True
9         True
10        True
11       False
12        True
13        True
14       False
15        True
16        True
17        True
18       False
19        True
20        True
21        True
22        True
23        True
24       False
25       False
26        True
27        True
28        True
29        True
         ...  
23517     True
23518     True
23519     True
23520     True
23521     True
23522     True
23523     True
23524     True
23525     True
23526     True
23527     True
23528     True
23529     True
23530     True
23531     True
23532     True
23533     True
23534     True
23535     True
23536     True
23537     True
23538     True
23539     True
23540     True
23541     True
23542     True
23543     True
23544     True
23545     True
23546     True
Length: 23547, dtype: bool

In [32]:
df.isnull().all(axis=1)
#this is to find all the missing values in a row

0        False
1        False
2        False
3        False
4        False
5        False
6        False
7        False
8        False
9        False
10       False
11       False
12       False
13       False
14       False
15       False
16       False
17       False
18       False
19       False
20       False
21       False
22       False
23       False
24       False
25       False
26       False
27       False
28       False
29       False
         ...  
23517    False
23518    False
23519    False
23520    False
23521    False
23522    False
23523    False
23524    False
23525    False
23526    False
23527    False
23528    False
23529    False
23530    False
23531    False
23532    False
23533    False
23534    False
23535    False
23536    False
23537    False
23538    False
23539    False
23540    False
23541    False
23542    False
23543    False
23544    False
23545    False
23546    False
Length: 23547, dtype: bool

In [33]:
df.isnull().all(axis=1).sum()

0

In [34]:
df.isnull().sum(axis=1)
#sum of missing values in each row

0        3
1        2
2        0
3        3
4        0
5        2
6        0
7        1
8        2
9        2
10       2
11       0
12       1
13       1
14       0
15       9
16       9
17       2
18       0
19       9
20       1
21       9
22       9
23       2
24       0
25       0
26       7
27       9
28       2
29       2
        ..
23517    2
23518    3
23519    3
23520    2
23521    1
23522    4
23523    3
23524    3
23525    1
23526    2
23527    1
23528    5
23529    4
23530    5
23531    9
23532    1
23533    5
23534    2
23535    3
23536    4
23537    2
23538    1
23539    2
23540    2
23541    1
23542    2
23543    8
23544    4
23545    1
23546    2
Length: 23547, dtype: int64

### treating missing values:
- there are two wasy to treat missing values
1. delete: Delete missing values
2. impute:
    - Imputing by a simple satistic: Replace the missing values by another value, commonly the mean, median, mode
    - Predictive techniques: Use statistical models such as K-NN, SVM etc., to predict and impute missing values.
- In generally, Imputation makes assumption about the missing values and replace missing values by arbitary numbers such as mean, median, mode etc,.. it should be used only when you are reasonably confident about the assumptions
- Otherwise, deletion is often safes and recommended. You may lose some data, but will not make any unreasonable assumptions.

Cautions: Always have a backup of the original data if you're deleting missing values

In [35]:
#to know the percentage of missing values in (col-wise)
round(100*(df.isnull().sum()/len(df.index)),2)  #-----> 2--> is decimals

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

In [36]:
#removing three cols
df=df.drop('BuildingArea',axis=1)
df=df.drop('YearBuilt',axis=1)
df=df.drop('CouncilArea',axis=1)

round(100*(df.isnull().sum()/len(df.index)),2)
######## This type of code should be only one time execution

# if any error do seperate

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

In [37]:
df=df.drop('CouncilArea',axis=1)
round(100*(df.isnull().sum()/len(df.index)),2)

KeyError: "['CouncilArea'] not found in axis"

In [38]:
df[df.isnull().sum(axis=1)>5]

Suburb              Address  Rooms Type      Price Method  \
15            Abbotsford     217 Langridge St      3    h  1000000.0      S   
16            Abbotsford      18a Mollison St      2    t   745000.0      S   
19            Abbotsford  403/609 Victoria St      2    u   542000.0      S   
21            Abbotsford    25/84 Trenerry Cr      2    u   760000.0     SP   
22            Abbotsford    106/119 Turner St      1    u   481000.0     SP   
27            Abbotsford    13/84 Trenerry Cr      1    u   500000.0      S   
33            Abbotsford     250 Langridge St      2    t   847000.0      S   
34            Abbotsford      16b Mollison St      2    h        NaN     PI   
45            Abbotsford    65/80 Trenerry Cr      1    u   480000.0      S   
46            Abbotsford  119/52 Nicholson St      1    u   423500.0      S   
52            Abbotsford    49/84 Trenerry Cr      2    h   700000.0      S   
60            Abbotsford    4/328 Johnston St      2    u        NaN     PI   
62            Abbotsford    82/84 Trenerry Cr      2    u   970000.0     SP   
65            Abbotsford    40/80 Trenerry Cr      3    h   927000.0      S   
74          Airport West          4/9 West Ct      2    t   420000.0     VB   
75          Airport West         1/36 York St      3    t   500000.0     VB   
76          Airport West          33 Bowes Av      5    h  1035000.0      S   
77          Airport West          89 Bowes Av      4    h   960000.0      S   
78          Airport West       45 Laurence Av      4    h   960000.0      S   
84          Airport West       81a Roberts Rd      3    h   680000.0     VB   
87          Airport West      3/24 Elstone Av      3    u   600000.0     VB   
89          Airport West         37a North St      3    t   760000.0     PI   
91          Airport West        1/69 Bowes Av      3    u   580000.0     VB   
92          Airport West       2/9 El Reno Cr      3    h   725000.0     PI   
93          Airport West       92 McNamara Av      4    h   758000.0      S   
95          Airport West       20 Creswell Av      3    h   970000.0      S   
97          Airport West          71 Bowes Av      4    h   985000.0      S   
99          Airport West       2/112 Parer Rd      2    u   509500.0     SP   
107         Airport West        7 McNamara Av      3    h   715000.0      S   
111         Airport West     1/41 Hillside Gr      2    t   460000.0     VB   
...                  ...                  ...    ...  ...        ...    ...   
23270            MacLeod          4 Tucker Wy      4    h   831000.0      S   
23271          Maidstone      88a Ballarat Rd      3    t   706000.0      S   
23290     Meadow Heights      22 Lightwood Cr      5    h   710000.0      S   
23291          Melbourne         505/1 Roy St      2    u        NaN     SP   
23292          Melbourne         505/1 Roy St      2    u        NaN      S   
23300            Mentone       20A Winsome St      3    t   905000.0      S   
23302          Mickleham         3 Selkirk Wy      2    h   335000.0      S   
23318       Moonee Ponds       55 Pattison St      2    h   700000.0     VB   
23333     Mount Waverley          19 Larch Cr      3    h        NaN     SN   
23337           Mulgrave         45 Carson St      3    t   826000.0      S   
23352          Northcote  2/241 Heidelberg Rd      2    t   770000.0     VB   
23363             Ormond        9a Bewdley St      4    t        NaN      S   
23384            Prahran         82 Murray St      2    h  1500000.0     VB   
23397          Reservoir        5a Kyneton Av      3    u   635000.0      S   
23398          Reservoir         13 Lawley St      4    h        NaN     PI   
23409           Richmond    10/9 Westbank Tce      2    u   410000.0     PI   
23412      Ringwood East     2/16 Bronhill Rd      4    t   850000.0     VB   
23418           Rowville       8 Lakeside Bvd      3    h   905000.0      S   
23424        Sandringham           4 Gipsy Wy      2    t        NaN

In [39]:
#count the num of rows having > 5 missing values
#use len(df.index)
len(df[df.isnull().sum(axis=1)>5].index)

4278

In [40]:
#4278 rows having more than 5 the missing  values
# calculate percentage
100*(len(df[df.isnull().sum(axis=1)>5].index)/len(df.index))

18.16791948018856

In [41]:
# retaining the rows having <=5 NaNs
df=df[df.isnull().sum(axis=1)<=5]

# look at the summary again
round(100*(df.isnull().sum()/len(df.index)),2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.71
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          1.05
Bathroom          1.07
Car               1.81
Landsize          9.65
Lattitude         0.13
Longtitude        0.13
Regionname        0.00
Propertycount     0.00
dtype: float64

In [42]:
# removing NaN price in rows
df= df[~np.isnan(df['Price'])]

round(100*(df.isnull().sum()/len(df.index)),2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.76
Landsize         9.83
Lattitude        0.15
Longtitude       0.15
Regionname       0.00
Propertycount    0.00
dtype: float64

In [43]:
len(df)

15086

In [44]:
df.dtypes

Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

In [45]:
df.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  15086.000000  1.508600e+04  15086.000000  15086.000000  14927.000000   
mean       2.932653  1.066695e+06     10.290753   3107.384330      2.913043   
std        0.963214  6.389331e+05      6.070538     94.360707      0.964641   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.400000e+05      6.100000   3046.000000      2.000000   
50%        3.000000  9.000000e+05      9.300000   3087.000000      3.000000   
75%        3.000000  1.320000e+06     13.100000   3149.750000      3.000000   
max       12.000000  9.000000e+06     48.100000   3977.000000     20.000000   

           Bathroom           Car       Landsize     Lattitude    Longtitude  \
count  14925.000000  14820.000000   13603.000000  15064.000000  15064.000000   
mean       1.538492      1.615520     558.116371    -37.809849    144.996338   
std        0.689311      0.955916    3987.326586      0.081152      0.106375   
min        0.000000      0.000000       0.000000    -38.182550    144.431810   
25%        1.000000      1.000000     176.500000    -37.858100    144.931193   
50%        1.000000      2.000000     440.000000    -37.803625    145.000920   
75%        2.000000      2.000000     651.000000    -37.756270    145.060000   
max        8.000000     10.000000  433014.000000    -37.408530    145.526350   

       Propertycount  
count   15086.000000  
mean     7491.225905  
std      4395.533579  
min       249.000000  
25%      4380.000000  
50%      6567.000000  
75%     10331.000000  
max     21650.000000

In [46]:
df['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

Notice that minimum is 0, maximum is 433014, the mean is 558 and median is 440(50%). There is a significant variation in the 25th and 75th percentile as well(176 to 651).

Thus imputing with mean/ median seems quite biased. and so we should remove the NaNs

In [47]:
# removing NaNs in  Landsize:
df= df[~np.isnan(df['Landsize'])]

round(100*(df.isnull().sum()/len(df.index)),2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.16
Longtitude       0.16
Regionname       0.00
Propertycount    0.00
dtype: float64

In [48]:
len(df[~np.isnan(df['Landsize'])])

13603

In [49]:
df

Suburb              Address  Rooms Type      Price Method  \
1             Abbotsford         85 Turner St      2    h  1480000.0      S   
2             Abbotsford      25 Bloomburg St      2    h  1035000.0      S   
4             Abbotsford         5 Charles St      3    h  1465000.0     SP   
5             Abbotsford     40 Federation La      3    h   850000.0     PI   
6             Abbotsford          55a Park St      4    h  1600000.0     VB   
10            Abbotsford       129 Charles St      2    h   941000.0      S   
11            Abbotsford         124 Yarra St      3    h  1876000.0      S   
14            Abbotsford        98 Charles St      2    h  1636000.0      S   
17            Abbotsford   6/241 Nicholson St      1    u   300000.0      S   
18            Abbotsford        10 Valiant St      2    h  1097000.0      S   
23            Abbotsford   411/8 Grosvenor St      2    u   700000.0     VB   
24            Abbotsford      40 Nicholson St      3    h  1350000.0     VB   
25            Abbotsford  123/56 Nicholson St      2    u   750000.0      S   
28            Abbotsford        45 William St      2    h  1172500.0      S   
29            Abbotsford   7/20 Abbotsford St      1    u   441000.0     SP   
30            Abbotsford        16 William St      2    h  1310000.0      S   
32            Abbotsford          42 Henry St      3    h  1200000.0      S   
35            Abbotsford          78 Yarra St      3    h  1176500.0      S   
36            Abbotsford     196 Nicholson St      3    h   955000.0      S   
37            Abbotsford        42 Valiant St      2    h   890000.0      S   
38            Abbotsford      3/72 Charles St      4    h  1330000.0     PI   
40            Abbotsford   13/11 Nicholson St      3    t   900000.0      S   
41            Abbotsford  138/56 Nicholson St      3    u  1090000.0      S   
42            Abbotsford   6/219 Nicholson St      2    u   500000.0      S   
43            Abbotsford       52a William St      2    h  1100000.0     PI   
44            Abbotsford           49 Park St      2    h  1315000.0      S   
48            Abbotsford   5/20 Abbotsford St      1    u   426000.0     SP   
49            Abbotsford     48 Abbotsford St      3    h  1447500.0     PI   
50            Abbotsford  116/56 Nicholson St      1    u   457000.0      S   
51            Abbotsford          159 Park St      2    h  1135000.0      S   
...                  ...                  ...    ...  ...        ...    ...   
23471     Sunshine North       5 Berkshire Rd      3    h   595000.0     SP   
23473      Sunshine West          24 Hilma St      4    h   682000.0      S   
23474      Sunshine West        16 Mudford St      3    h   640000.0      S   
23475       Surrey Hills         20 Albert Cr      4    h  2720000.0      S   
23476       Surrey Hills         46 Durham Rd      3    h  1715000.0      S   
23479       Surrey Hills             3 Oak St      4    h  3100000.0     VB   
23482            Tarneit        27 McMahon Cr      3    h   350000.0     VB   
23488        Templestowe       10 Hathaway Cl      4    h  1753000.0      S   
23490  Templestowe Lower          20 Airds Rd      4    h  1745000.0      S   
23491  Templestowe Lower       16 Benambra Dr      3    h  1190000.0      S   
23492  Templestowe Lower       50 Caroline Dr      5    h  1400000.0      S   
23493  Templestowe Lower            7 Dove Ct      4    h  1288000.0      S   
23497          Thornbury      7 Ballantyne St      3    h  1450000.0      S   
23501          Thornbury         201 Gooch St      3    h  1271000.0      S   
23512        Tullamarine         7 Londrew Ct      3    h   540000.0      S   
23514            Vermont       73 Nurlendi Rd      4    h  1263500.0      S   
23518      Vermont South        9 Winswood Cl      4    h  1250000.0     PI   
23519           Viewbank        149 Graham Rd      5    h  1316000.0     SP   
23521           Wantirna   16 chesterfield Ct      4    h   951000.0

In [52]:
df['Lattitude'].describe()

count    13581.000000
mean       -37.809204
std          0.079257
min        -38.182550
25%        -37.856820
50%        -37.802360
75%        -37.756400
max        -37.408530
Name: Lattitude, dtype: float64

In [54]:
df['Lattitude'].fillna(df['Lattitude'].mean(),inplace=True)

In [55]:
df['Lattitude'].mean()

-37.80920350636921

In [56]:
df.dtypes

Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

In [58]:
df['Longtitude'].describe()

count    13581.000000
mean       144.995221
std          0.103913
min        144.431810
25%        144.929600
50%        145.000100
75%        145.058320
max        145.526350
Name: Longtitude, dtype: float64

In [59]:
df['Longtitude'].fillna(df['Longtitude'].mean(),inplace=True)

In [60]:
df

Suburb              Address  Rooms Type      Price Method  \
1             Abbotsford         85 Turner St      2    h  1480000.0      S   
2             Abbotsford      25 Bloomburg St      2    h  1035000.0      S   
4             Abbotsford         5 Charles St      3    h  1465000.0     SP   
5             Abbotsford     40 Federation La      3    h   850000.0     PI   
6             Abbotsford          55a Park St      4    h  1600000.0     VB   
10            Abbotsford       129 Charles St      2    h   941000.0      S   
11            Abbotsford         124 Yarra St      3    h  1876000.0      S   
14            Abbotsford        98 Charles St      2    h  1636000.0      S   
17            Abbotsford   6/241 Nicholson St      1    u   300000.0      S   
18            Abbotsford        10 Valiant St      2    h  1097000.0      S   
23            Abbotsford   411/8 Grosvenor St      2    u   700000.0     VB   
24            Abbotsford      40 Nicholson St      3    h  1350000.0     VB   
25            Abbotsford  123/56 Nicholson St      2    u   750000.0      S   
28            Abbotsford        45 William St      2    h  1172500.0      S   
29            Abbotsford   7/20 Abbotsford St      1    u   441000.0     SP   
30            Abbotsford        16 William St      2    h  1310000.0      S   
32            Abbotsford          42 Henry St      3    h  1200000.0      S   
35            Abbotsford          78 Yarra St      3    h  1176500.0      S   
36            Abbotsford     196 Nicholson St      3    h   955000.0      S   
37            Abbotsford        42 Valiant St      2    h   890000.0      S   
38            Abbotsford      3/72 Charles St      4    h  1330000.0     PI   
40            Abbotsford   13/11 Nicholson St      3    t   900000.0      S   
41            Abbotsford  138/56 Nicholson St      3    u  1090000.0      S   
42            Abbotsford   6/219 Nicholson St      2    u   500000.0      S   
43            Abbotsford       52a William St      2    h  1100000.0     PI   
44            Abbotsford           49 Park St      2    h  1315000.0      S   
48            Abbotsford   5/20 Abbotsford St      1    u   426000.0     SP   
49            Abbotsford     48 Abbotsford St      3    h  1447500.0     PI   
50            Abbotsford  116/56 Nicholson St      1    u   457000.0      S   
51            Abbotsford          159 Park St      2    h  1135000.0      S   
...                  ...                  ...    ...  ...        ...    ...   
23471     Sunshine North       5 Berkshire Rd      3    h   595000.0     SP   
23473      Sunshine West          24 Hilma St      4    h   682000.0      S   
23474      Sunshine West        16 Mudford St      3    h   640000.0      S   
23475       Surrey Hills         20 Albert Cr      4    h  2720000.0      S   
23476       Surrey Hills         46 Durham Rd      3    h  1715000.0      S   
23479       Surrey Hills             3 Oak St      4    h  3100000.0     VB   
23482            Tarneit        27 McMahon Cr      3    h   350000.0     VB   
23488        Templestowe       10 Hathaway Cl      4    h  1753000.0      S   
23490  Templestowe Lower          20 Airds Rd      4    h  1745000.0      S   
23491  Templestowe Lower       16 Benambra Dr      3    h  1190000.0      S   
23492  Templestowe Lower       50 Caroline Dr      5    h  1400000.0      S   
23493  Templestowe Lower            7 Dove Ct      4    h  1288000.0      S   
23497          Thornbury      7 Ballantyne St      3    h  1450000.0      S   
23501          Thornbury         201 Gooch St      3    h  1271000.0      S   
23512        Tullamarine         7 Londrew Ct      3    h   540000.0      S   
23514            Vermont       73 Nurlendi Rd      4    h  1263500.0      S   
23518      Vermont South        9 Winswood Cl      4    h  1250000.0     PI   
23519           Viewbank        149 Graham Rd      5    h  1316000.0     SP   
23521           Wantirna   16 chesterfield Ct      4    h   951000.0

In [71]:
df['Car'].describe()

count     13540.0
unique       11.0
top           2.0
freq       5606.0
Name: Car, dtype: float64

In [63]:
#converting to type 'Category'
df['Car']=df['Car'].astype('category')

In [66]:
df['Car'].head()

1    1.0
2    0.0
4    0.0
5    1.0
6    2.0
Name: Car, dtype: category
Categories (11, float64): [0.0, 1.0, 2.0, 3.0, ..., 7.0, 8.0, 9.0, 10.0]

In [67]:
df['Car'].unique()

[1.0, 0.0, 2.0, 6.0, 5.0, ..., 8.0, 7.0, 9.0, 10.0, NaN]
Length: 12
Categories (11, float64): [1.0, 0.0, 2.0, 6.0, ..., 8.0, 7.0, 9.0, 10.0]

In [69]:

df['Car'].nunique()

11

In [70]:
df['Car'].value_counts()

2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [72]:
df['Car'].fillna(2.0,inplace=True)

In [73]:
df['Bathroom'].describe()

count    13602.000000
mean         1.534921
std          0.691834
min          0.000000
25%          1.000000
50%          1.000000
75%          2.000000
max          8.000000
Name: Bathroom, dtype: float64

In [74]:
#converting to type 'Category'
df['Bathroom']=df['Bathroom'].astype('category')

In [75]:
df['Bathroom'].head()

1    1.0
2    1.0
4    2.0
5    2.0
6    1.0
Name: Bathroom, dtype: category
Categories (9, float64): [0.0, 1.0, 2.0, 3.0, ..., 5.0, 6.0, 7.0, 8.0]

In [76]:
df['Bathroom'].unique()

[1.0, 2.0, 3.0, 4.0, 0.0, 7.0, 5.0, 6.0, 8.0, NaN]
Categories (9, float64): [1.0, 2.0, 3.0, 4.0, ..., 7.0, 5.0, 6.0, 8.0]

In [77]:
df['Bathroom'].nunique()

9

In [79]:
df['Bathroom'].value_counts()

1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
8.0       2
7.0       2
Name: Bathroom, dtype: int64

In [80]:
df['Bathroom'].fillna(1.0,inplace=True)